provider/equinix/environ.go

```diff
@@ -0,0 +1,784 @@
+		for _, os := range oss {
+			if !isDistroSupported(os, ic) {
+				continue
+			}
+
+			for _, p := range os.ProvisionableOn {
+				if p == it.Name {
+					image := &imagemetadata.ImageMetadata{
+						Id:   os.Slug,
+						Arch: getArchitectureFromPlan(p),
+					}
+					suitableImages = append(suitableImages, image)
+				}
+			}
+		}
+	}
+
+	images := instances.ImageMetadataToImages(suitableImages)
+	return instances.FindInstanceSpec(images, ic, instanceTypes)
+}
+
+func (e *environ) finishInstanceConfig(args *environs.StartInstanceParams, spec *instances.InstanceSpec) error {
+	matchingTools, err := args.Tools.Match(tools.Filter{Arch: spec.Image.Arch})
+	if err != nil {
+		return errors.Errorf("chosen architecture %v for image %q not present in %v",
+			spec.Image.Arch, spec.Image.Id, args.Tools.Arches())
+	}
+
+	if spec.InstanceType.Deprecated {
+		logger.Warningf("deprecated instance type specified: %s", spec.InstanceType.Name)
+	}
+
+	if err := args.InstanceConfig.SetTools(matchingTools); err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := instancecfg.FinishInstanceConfig(args.InstanceConfig, e.Config()); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+var ErrDeviceProvisioningFailed = errors.New("device provisioning failed")
+
+// waitDeviceActive is a function capable of figuring out when a Equinix Metal
+// device is active
+func waitDeviceActive(ctx context.ProviderCallContext, c *packngo.Client, id string) (*packngo.Device, error) {
+	var d *packngo.Device
+	err := retry.Call(retry.CallArgs{
+		Func: func() error {
+			var err error
+			d, _, err = c.Devices.Get(id, nil)
+			if err != nil {
+				return err
+			}
+			if d.State == "active" {
+				return nil
+			}
+			if d.State == "failed" {
+				return errr.Wrap(ErrDeviceProvisioningFailed, fmt.Sprintf("device %s provisioning failed", id))
+			}
+			return fmt.Errorf("device in not in active state yet")
+		},
+		IsFatalError: func(err error) bool {
+			if errr.Is(err, ErrDeviceProvisioningFailed) {
+				return true
+			}
+			select {
+			case <-ctx.Dying():
+				return true // treat error as fatal because we are asked to abort
+			default:
+				return common.IsCredentialNotValid(err)
+			}
+		},
+		Attempts: 180,
+		Delay:    5 * time.Second,
+		Clock:    clock.WallClock,
+	})
+
+	if d != nil {
+		return d, nil
+	}
+
+	return nil, err
+}
+
+// Helper function to get supported OS version
+func isDistroSupported(os packngo.OS, ic *instances.InstanceConstraint) bool {
+	switch os.Distro {
+	case "ubuntu":
+		series, err := series.VersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	case "centos":
+		series, err := series.CentOSVersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	case "windows":
+		series, err := series.WindowsVersionSeries(os.Version)
+		if err != nil || ic.Series != series {
+			return false
+		}
+	default:
+		return false
+	}
+
+	return true
```

Comment:
This should return false and then you can drop the default.

Reply:
I don't think the proposal you have done keeps compatibility with how the current code works.
I wrote a test that fails if I changed the code as you suggested

Reply:
You're right, can we keep the test though?